### ISSUE TITLE
[Inconsistency] Auth key is derived by hashing the public key WITHOUT salt and LCS serialization

### ISSUE THREAD

**Issue by valerini** [libra/libra#6549]
# 🐛 Inconsistency/Bug
Not a real bug, just an inconsistency that we might want to fix as it will be harder to fix later.

It was intended to hash structs only serializable with LCS, in which case the "canonical" hasher can be automatically derived (see [the specification for Domain Separation](https://github.com/libra/libra/tree/master/specifications/crypto#domain-separation)):
- it will serialize the struct with LCS (Libra Canonical Serializer),
- it will prepend the salt derived from the struct's name,
- finally it will hash the result with sha3-256.

The authentication key is not derived with the hashing procedure outlined above: see the implementation of `AuthenticationKey` at [types/src/transaction/authenticator.rs:164](https://github.com/libra/libra/blob/10dc3cb9e645eec840139508b9d813ea6dc7108e/types/src/transaction/authenticator.rs#L164). The input to the hash is the concatenation of the bytes of the public key and the id of scheme, this encoding is collision free now (as we only have two signature schemes - EdDSA and Multi-EdDSA), but it is likely not to remain such if we add more signature schemes.

 Note that also the validator owner's auth key is not derived via "canonical" hashing: (see [utils.rs:144](https://github.com/libra/libra/blob/10dc3cb9e645eec840139508b9d813ea6dc7108e/config/src/utils.rs#L14)).

Please leave your thoughts on whether this should be fixed now in which case it will be a breaking change.

**Comment by mimoo** [comment]
yeah there's a number of places where we don't use the hasher, thomas pornin was also pointing at hashing internal nodes in the sparse merkle tree.

**Comment by valerini** [comment]
@mimoo, does not seem to me to be obviously a problem. The only thing that's worrisome is that it will not be possible to change later on. And it seems, new problems will appear for the auth-key in particular if/when we add more signature schemes (I think for internal Merkle nodes it is not such a big of an issue).

**Comment by n4ss** [comment]
That'd be a breaking change on the account creation side by the way, right?

**Comment by sblackshear** [comment]
> The input to the hash is the concatenation of the bytes of the public key and the id of scheme, this encoding is collision free now (as we only have two signature schemes - EdDSA and Multi-EdDSA), but it is likely not to remain such if we add more signature schemes.

I'm not sure I understand this part. If we add a new signature ID for each new scheme, is that sufficient to avoid collisions?

**Comment by valerini** [comment]
@sblackshear Sorry, my bad, adding the id is enough to avoid collisions indeed.
